MCB-Nucleo/Core/Src/freertos.c

```diff
@@ -0,0 +1,454 @@
+{
+  /* USER CODE BEGIN startGetCANVelocity */
+	uint8_t CAN_message[8];
+	FloatBytes velocity;
+	/* Infinite loop */
+	for(;;)
+	{
+		if (HAL_CAN_GetRxFifoFillLevel(&hcan, CAN_RX_FIFO0))
+		{
+			// there are multiple CAN IDs being passed through the filter, check if the message is the SOC
+			HAL_CAN_GetRxMessage(&hcan, CAN_RX_FIFO0, &can_rx_header, CAN_message);
+			if (can_rx_header.StdId == 0x503)
+			{
+				for(int i= 0; i < 4; i++)
+				{
+					velocity.bytes[i] = CAN_message[i+4]; // Vechicle Velocity is stored in bits 32-63.
+				}
+				velocity_of_car = velocity.float_value;
+			}
+		}
+		osDelay(GET_CAN_VELOCITY_DELAY);
+	}
+  /* USER CODE END startGetCANVelocity */
+}
+
+/* USER CODE BEGIN Header_StartSetCANBatterySO */
+/**
+* @brief Function implementing the getCANBatterySO thread.
+* @param argument: Not used
+* @retval None
+*/
+/* USER CODE END Header_StartSetCANBatterySO */
+void StartSetCANBatterySO(void *argument)
+{
+  /* USER CODE BEGIN StartSetCANBatterySO */
+	uint8_t battery_msg_data[8];
+	/* Infinite loop */
+	for(;;)
+	{
+		if (HAL_CAN_GetRxFifoFillLevel(&hcan, CAN_RX_FIFO0))
+		{
+			// there are multiple CAN IDs being passed through the filter, check if the message is the SOC
+			HAL_CAN_GetRxMessage(&hcan, CAN_RX_FIFO0, &can_rx_header, battery_msg_data);
+			if (can_rx_header.StdId == 0x626)
+			{
+				// if the battery SOC is out of range, assume it is at 100% as a safety measure
+				if (battery_msg_data[0] < 0 || battery_msg_data[0] > 100)
+					battery_soc = 100;
+				else
+					battery_soc = battery_msg_data[0];
+			}
+
+	  		osDelay(READ_BATTERY_SOC_DELAY);
+		}
+	}
+  /* USER CODE END StartSetCANBatterySO */
+}
+
+/* USER CODE BEGIN Header_StartUpdateState */
+/**
+* @brief Function implementing the updateState thread.
+* @param argument: Not used
+* @retval None
+*/
+/* USER CODE END Header_StartUpdateState */
+void StartUpdateState(void *argument)
+{
+  /* USER CODE BEGIN StartUpdateState */
+  /* Infinite loop */
+  for(;;)
+  {
+	  if (event_flags.park_enabled && (state == IDLE || state == PARK))
+		  state = PARK;
+	  else if (event_flags.mech_brake_pressed)
+		  state = IDLE;
+	  else if (event_flags.regen_pressed && event_flags.charge_under_threshold)
+	  	  state = REGEN;
+	  else if (event_flags.cruise_enabled)
+	      state = CRUISE;
+	  else if (event_flags.reverse_enabled && event_flags.velocity_under_threshold && state != PARK)
+		  state = REVERSE;
+	  else if (event_flags.throttle_pressed)
+	      state = DRIVE;
+	  else
+	  	  state = IDLE;
+
+	  osDelay(EVENT_FLAG_UPDATE_DELAY);
+  }
+  /* USER CODE END StartUpdateState */
+}
+
+/* Private application code --------------------------------------------------*/
+/* USER CODE BEGIN Application */
+void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
+{
+	if(GPIO_Pin == BTN_CRUISE_TOGGLE_Pin)
+	{
+
+		if(state == DRIVE || state == CRUISE)
+		{
+			event_flags.cruise_enabled = !event_flags.cruise_enabled;
+			cruise_velocity = velocity_of_car;
+		}
```

Comment:
Outside code. Make sure this is specced that the cruise button also turns off cruise - if that is the desired behavior.